[go] Bump images, dependencies and versions to go 1.24.0

cpanato opened the pull request:
#### What type of PR is this?

/kind feature
/hold

#### What this PR does / why we need it:

- Bump images, dependencies and versions to go 1.24.0

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/release/issues/3772

/assign @MadhavJivrajani @liggitt @puerco @xmudrii  @Verolop  
cc @kubernetes/release-managers 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Kubernetes is now built with go 1.24.0
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

cpanato committed:
Bump images, dependencies and versions to go 1.24.0

Signed-off-by: cpanato <ctadeu@gmail.com>

cpanato commented:
/assign @MadhavJivrajani @liggitt @puerco @xmudrii @Verolop

liggitt commented:
I was *really* hoping sig-storage / sig-windows would complete the sweep and update of windows symlink handling to complete our transition to go 1.23 before we moved on to go 1.24 :-/

Resolution of https://github.com/kubernetes/kubernetes/issues/129084 should be higher priority than go 1.24, in my opinion (which is tricky since the folks working on that are non-overlapping with the ones updating to go 1.24)

cpanato commented:
@liggitt @MadhavJivrajani @dims ready

dims commented:
/approve
/lgtm

left a note for @mansikulkarni96 on that other issue https://github.com/kubernetes/kubernetes/issues/129084#issuecomment-2674616104

liggitt commented:
All the go mod files should change to go 1.24, including the `godebug default` lines, to make sure tests actually work with go 1.24 behavior. That includes dropping the winsymlink workaround that we're currently blocked on from the storage and windows sweep.

liggitt commented:
https://github.com/kubernetes/kubernetes/issues/129084 is now resolved

Update the go.mod and go.work files to 1.24.0, and then this should be set, assuming it passes CI

dims commented:
/retest

dims commented:
/test all

dims commented:
/test pull-kubernetes-linter-hints 
/test pull-kubernetes-verify

ameukam commented:
/retest

cpanato committed:
bump go.mod to set min go1.24

Signed-off-by: cpanato <ctadeu@gmail.com>

dims commented:
> @dims @liggitt we might need to update the job spec for pull-publishing-bot-validate it is using       image: public.ecr.aws/docker/library/golang:1.23 if you are fine if that i will open a pr

@cpanato yes please!

liggitt commented:
the unit test failures look legitimate/relevant

@enj what needs to happen to get the kms job updated to pass? Is there a way to make that work automatically with .go-version?

ameukam commented:
Looks like we need to bump https://github.com/kubernetes/kubernetes/blob/4b12e89d0d1b8eefbe94f615ede351ccb9501164/staging/src/k8s.io/kms/internal/plugins/_mock/Dockerfile#L15 ?
In the past: https://github.com/kubernetes/kubernetes/pull/127278

enj commented on staging/src/k8s.io/kms/internal/plugins/_mock/go.mod:
```diff
@@ -1,6 +1,6 @@
-go 1.23.0
+go 1.24.0
 
 godebug default=go1.23
```
Also need to update this?

dims commented:
@cpanato 3 more spots where we have `1.23`

```
❯ rg go1.23 | grep go.mod
hack/tools/go.mod:godebug default=go1.23
staging/src/k8s.io/kms/internal/plugins/_mock/go.mod:godebug default=go1.23
staging/src/k8s.io/code-generator/examples/go.mod:godebug default=go1.23
```

liggitt committed:
Switch to private instances of rand for seeding for tests

liggitt commented:
last commit of https://github.com/liggitt/kubernetes/commits/update-main-go124/ fixes the unit test rand.Seed issues if you want to cherry-pick it in

liggitt commented:
> > what needs to happen to get the kms job updated to pass?
> 
> I think the latest changes to `go.work` should fix it.
> 
> > Is there a way to make that work automatically with .go-version?
> 
> Do we have an existing example of using that file to template the go version into go.mod/go.work/docker files?

Add an arg / var to the dockerfile:
https://github.com/kubernetes/kubernetes/blob/4b12e89d0d1b8eefbe94f615ede351ccb9501164/build/build-image/Dockerfile/#L16-L19

Pass a --builder-arg to the docker build invocation:
https://github.com/kubernetes/kubernetes/blob/4b12e89d0d1b8eefbe94f615ede351ccb9501164/build/common.sh#L392

Calculate the builder-arg however you want... probably from .go-version in https://github.com/kubernetes/kubernetes/blob/master/test/e2e/testing-manifests/auth/encrypt/run-e2e.sh#L33 or something

cpanato commented:
> last commit of https://github.com/liggitt/kubernetes/commits/update-main-go124/ fixes the unit test rand.Seed issues if you want to cherry-pick it in

cherry picked

liggitt commented:
squash the "Bump images, dependencies and versions to ..." commits to a single commit, then lgtm once CI is happy 🎉

siyuanfoundation commented:
/triage accepted

dims commented:
/test pull-publishing-bot-validate

dims commented:
wondering why `pull-kubernetes-verify` and `pull-kubernetes-linter-hints` are taking too much time ...

liggitt commented:
> Test started today at 10:44 AM is still running after 6h45m30s. (more info)
> ...
> +++ Running case: verify.golangci-lint 

hmm

liggitt commented:
maybe https://github.com/golangci/golangci-lint/issues/5225 is relevant? 

> EDIT: since v1.64 golangci-lint supports go1.24 if compiled with go1.24 🎉

we're on `github.com/golangci/golangci-lint v1.56.2`

liggitt committed:
Bump hack/tools to github.com/golangci/golangci-lint 1.64.5 for go 1.24

liggitt commented:
> we're on github.com/golangci/golangci-lint v1.56.2 ... I bet we need to bump github.com/golangci/golangci-lint to at least 1.26 in hack/tools/go.mod
> 
> Let's do that in a distinct commit.

Last commit of https://github.com/liggitt/kubernetes/commits/update-main-go124/ does that if you want to pick it in

cpanato commented:
@liggitt @dims saw some lint errors, I can fix those in this PR if needed

dims commented:
@pohly what do you want to do here for the linter issues?

pohly commented:
Some of these must might be new checks. As we want to do the same bump in release branches and shouldn't make sweeping code changes in the release branches, we should disable these checks for now. The once this PR is merged, we can re-evaluate whether some of them should be fixed and enforced. Agreed?

If yes, then I can prepare a PR against @cpanato's branch.

dims commented:
> we should disable these checks for now. The once this PR is merged, we can re-evaluate whether some of them should be fixed and enforced. Agreed?

Agree and thanks @pohly

liggitt commented:
We should double check that the new lint errors are actually new checks. If they are, disabling for now sounds good (ideally squashed into the commit that bumps golangci-lint for easy picking to release branches), and then selectively re-enabling and fixing in master for 1.33+ in a follow-up.

pohly committed:
golangci-lint: temporarily suppress known issues in the code base

These suppressions are necessary to make golangci-lint 1.64 pass with the
current code base. This change is meant to be backported to release
branches. On master, we may want to revert some of it together with fixing the
findings.

pohly commented:
@cpanato: your repo doesn't seem to accept PRs, so here's a patch instead:

```patch
From 3ef037f907df4858cb51f4ca1a87dd8a5303fb02 Mon Sep 17 00:00:00 2001
From: Patrick Ohly <patrick.ohly@intel.com>
Date: Wed, 26 Feb 2025 15:52:08 +0100
Subject: [PATCH 1/1] golangci-lint: temporarily suppress known issues in the
 code base

These suppressions are necessary to make golangci-lint 1.64 pass with the
current code base. This change is meant to be backported to release
branches. On master, we may want to revert some of it together with fixing the
findings.
---
 hack/golangci.yaml    | 18 ++++++++++++++++++
 hack/golangci.yaml.in | 21 +++++++++++++++++++++
 2 files changed, 39 insertions(+)

diff --git a/hack/golangci.yaml b/hack/golangci.yaml
index eebd23a700f..f7c834116f3 100644
--- a/hack/golangci.yaml
+++ b/hack/golangci.yaml
@@ -73,6 +73,24 @@ issues:
        - ginkgolinter
       text: use a function call in (Eventually|Consistently)
 
+    # Some of these seem legitimate, maybe better fix code (https://github.com/kubernetes/kubernetes/issues/130449).
+
+    - linters:
+        - govet
+      text: "lostcancel|printf"
+
+    - linters:
+        - ginkgolinter
+      text: "wrong error assertion. Consider using `gomega.(Eventually|Consistently)|wrong comparison assertion|wrong length assertion"
+
+    - linters:
+        - testifylint
+      text: "encoded-compare|error-nil|formatter"
+
+    - linters:
+        - gocritic
+      text: "append result not assigned to the same slice|put a space between `//` and comment text|sloppyLen|elseif|should rewrite switch statement to if statement|regexpMust|wrapperFunc: use strings.ReplaceAll|singleCaseSwitch|deprecatedComment|exitAfterDefer|captLocal|unlambda|underef|unslice|valSwap|typeSwitchVar"
+
     # https://github.com/kubernetes/kubernetes/issues/117288#issuecomment-1507012435
     - linters:
         - gocritic
diff --git a/hack/golangci.yaml.in b/hack/golangci.yaml.in
index 08b556a08be..6c6ed5e99c6 100644
--- a/hack/golangci.yaml.in
+++ b/hack/golangci.yaml.in
@@ -76,6 +76,27 @@ issues:
        - ginkgolinter
       text: use a function call in (Eventually|Consistently)
 
+    {{- if .Base}}
+
+    # Some of these seem legitimate, maybe better fix code (https://github.com/kubernetes/kubernetes/issues/130449).
+
+    - linters:
+        - govet
+      text: "lostcancel|printf"
+
+    - linters:
+        - ginkgolinter
+      text: "wrong error assertion. Consider using `gomega.(Eventually|Consistently)|wrong comparison assertion|wrong length assertion"
+
+    - linters:
+        - testifylint
+      text: "encoded-compare|error-nil|formatter"
+
+    - linters:
+        - gocritic
+      text: "append result not assigned to the same slice|put a space between `//` and comment text|sloppyLen|elseif|should rewrite switch statement to if statement|regexpMust|wrapperFunc: use strings.ReplaceAll|singleCaseSwitch|deprecatedComment|exitAfterDefer|captLocal|unlambda|underef|unslice|valSwap|typeSwitchVar"
+    {{- end}}
+
     # https://github.com/kubernetes/kubernetes/issues/117288#issuecomment-1507012435
     - linters:
         - gocritic
-- 
2.39.2

```

pohly commented:
> If they're not new checks which are newly failing with go 1.24, we should figure out why and decide whether to fix or disable the checks case-by-case.

Does it matter why we now get these issues reported? Isn't it more important whether they are real bugs that must be fixed?

I think govet "lostcancel" has been around for a while. It now seems to find a few additional occurrences. This is definitely something that I would fix, but even then a follow-up PR may be better? It might be worth fixing on master, but not worth backporting.

I most confused why gocritic now has new findings. But at the same time those seem least worthy of fixing because they are opinionated ("append result not assigned to the same slice") or cosmetic ("put a space between `//` and comment text").

pohly commented:
> you cna push to your branch and then i can cherry pick

https://github.com/kubernetes/kubernetes/compare/master...pohly:kubernetes:update-main-go124?expand=1

liggitt commented:
> > If they're not new checks which are newly failing with go 1.24, we should figure out why and decide whether to fix or disable the checks case-by-case.
> 
> Does it matter why we now get these issues reported? Isn't it more important whether they are real bugs that must be fixed?

I completely agree we should fix real issues. Understanding why they're newly reported would help us know if they are *new* issues related to the go 1.24 bump in some way that we should fix *with* this bump, rather than in a follow-up.

liggitt commented:
/lgtm
/approve

once CI is happy

pohly commented:
> Understanding why they're newly reported would help us know if they are new issues related to the go 1.24 bump in some way that we should fix with this bump, rather than in a follow-up.

Ack.

We can ignore gocritic. All of the reported issue are stylistic and not regressions caused by Go 1.24.

We can ignore testifylint and ginkgolinter. I'm pretty sure these are enhanced checks and only occur in tests.

govet's printf check is new. We knew that the Go team decided to enable that by default and fixed some occurrences, but since then this one got added in 9b6ef250fc01f6f2a645382d1e9397435e069265:

```
ERROR: test/e2e/node/mount_propagation.go:200:18: printf: non-constant format string in call to k8s.io/kubernetes/test/e2e/framework.Logf (govet)
ERROR: 		framework.Logf(enterKubeletMountNS)
```

That leaves govet's lostcancel:
```
ERROR: staging/src/k8s.io/client-go/tools/cache/reflector_test.go:999:2: lostcancel: the cancel function is not used on all paths (possible context leak) (govet)
ERROR: 	cancelCtx, cancel = context.WithCancelCause(ctx)
ERROR: 	^
ERROR: staging/src/k8s.io/client-go/tools/cache/reflector_test.go:1004:1: lostcancel: this return statement may be reached without using the cancel var defined on line 999 (govet)
ERROR: }
ERROR: ^
ERROR: staging/src/k8s.io/client-go/tools/cache/reflector_test.go:1050:2: lostcancel: the cancel function is not used on all paths (possible context leak) (govet)
ERROR: 	cancelCtx, cancel = context.WithCancelCause(ctx)
ERROR: 	^
ERROR: staging/src/k8s.io/client-go/tools/cache/reflector_test.go:1062:1: lostcancel: this return statement may be reached without using the cancel var defined on line 1050 (govet)
ERROR: }
ERROR: ^
ERROR: staging/src/k8s.io/client-go/tools/cache/reflector_test.go:1115:2: lostcancel: the cancel function is not used on all paths (possible context leak) (govet)
ERROR: 	cancelCtx, cancel = context.WithCancelCause(ctx)
ERROR: 	^
ERROR: staging/src/k8s.io/client-go/tools/cache/reflector_test.go:1127:1: lostcancel: this return statement may be reached without using the cancel var defined on line 1115 (govet)
ERROR: }
ERROR: ^
ERROR: staging/src/k8s.io/client-go/tools/cache/reflector_test.go:1187:2: lostcancel: the cancel function is not used on all paths (possible context leak) (govet)
ERROR: 	cancelCtx, cancel = context.WithCancelCause(ctx)
ERROR: 	^
ERROR: staging/src/k8s.io/client-go/tools/cache/reflector_test.go:1201:1: lostcancel: this return statement may be reached without using the cancel var defined on line 1187 (govet)
ERROR: }
ERROR: ^
ERROR: staging/src/k8s.io/client-go/tools/cache/reflector_test.go:1265:3: lostcancel: the cancel function is not used on all paths (possible context leak) (govet)
ERROR: 		cancelCtx, cancel = context.WithCancelCause(ctx)
ERROR: 		^
ERROR: staging/src/k8s.io/client-go/tools/cache/reflector_test.go:1275:1: lostcancel: this return statement may be reached without using the cancel var defined on line 1265 (govet)
ERROR: }
ERROR: ^
```

I think that is because of added support for `context.WithCancelCause`. If I replace that with `context.WithCancel` on current master without this PR, I get the same warning:
```
ERROR: staging/src/k8s.io/client-go/tools/cache/reflector_test.go:1265:3: lostcancel: the cancel function is not used on all paths (possible context leak) (govet)
ERROR: 		cancelCtx, cancel = context.WithCancel(ctx)
ERROR: 		^
ERROR: staging/src/k8s.io/client-go/tools/cache/reflector_test.go:1275:1: lostcancel: this return statement may be reached without using the cancel var defined on line 1265 (govet)
ERROR: }
ERROR: ^
```

However, I'm not seeing anything in the Go release notes about this added support. Perhaps it was considered too minor to be mentioned.

rajibmitra commented:
@pohly maybe we can try adding  defer cancel() after each context.

pohly commented:
> maybe we can try adding defer cancel() after each context.

Yes, that'll be in a follow-up.

liggitt commented:
https://storage.googleapis.com/k8s-triage/index.html?pr=1&text=failed%20to%20create%20fsnotify%20watcher%3A%20too%20many%20open%20files

pre-existing e2e flake

/retest

liggitt commented:
hmmm... same failure

> Kubernetes e2e suite: [It] [sig-cli] Kubectl client Simple pod should contain last line of the log
>
> { failed [FAILED] Expected     <string>: failed to create fsnotify watcher: too many open files to contain substring     <string>: EOF

rajibmitra commented:
> hmmm... same failure
> 
> > Kubernetes e2e suite: [It] [sig-cli] Kubectl client Simple pod should contain last line of the log
> > { failed [FAILED] Expected
> > `<string>: failed to create fsnotify watcher: too many open files`
> > to contain substring
> > `<string>: EOF`

@liggitt is there any way we could increase the ulimit ?

liggitt commented:
> @liggitt is there any way we could increase the ulimit ?

I'm more interested in knowing why go 1.24 exacerbates the failure than working around it

rajibmitra commented:
> > @liggitt is there any way we could increase the ulimit ?
> 
> I'm more interested in knowing why go 1.24 exacerbates the failure than working around it

Imho, once we have this in place https://github.com/kubernetes/kubernetes/pull/129688#issuecomment-2685491110 , go 1.24 should stop complaining, seems like latest one is stricter.

liggitt commented:
> Imho, once we have this in place [#129688 (comment)](https://github.com/kubernetes/kubernetes/pull/129688#issuecomment-2685491110) , go 1.24 should stop complaining, seems like latest one is stricter.

hrmm, but were any of the flagged locations relevant to the *e2e* test that's failing? I thought those were mostly in unit test locations we didn't care about

liggitt commented:
/hold
/retest

rajibmitra commented:
> > Imho, once we have this in place [#129688 (comment)](https://github.com/kubernetes/kubernetes/pull/129688#issuecomment-2685491110) , go 1.24 should stop complaining, seems like latest one is stricter.
> 
> hrmm, but were any of the flagged locations relevant to the _e2e_ test that's failing? I thought those were mostly in unit test locations we didn't care about

not sure of the locations, but i can quickly raise a PR that fixes the context test cases against master branch.

liggitt commented:
yeah, the failing test is regularly failing in CI periodics already - https://storage.googleapis.com/k8s-triage/index.html?text=failed%20to%20create%20fsnotify%20watcher%3A%20too%20many%20open%20files

we need an open issue for the test / component owner to resolve that pretty quickly

/hold cancel

liggitt commented:
/retest

cpanato commented:
nice!

BenTheElder commented:
For next time we're also working on cutting down the number of independent locations that need updating.

https://github.com/kubernetes/kubernetes/pull/130453

https://kubernetes.slack.com/archives/CHGFYJVAN/p1740625989283549 / https://github.com/kubernetes/test-infra/pull/34411#issuecomment-2683325052

I think most of our tests / builds / ... can just consume the .go-version / GOTOOLCHAIN / .. logic and kube-cross image

Then it's just an optimization to update the CI images to have the desired version pre-installed without having to fetch it via gimme / GOTOOLCHAIN, we shouldn't need to block on PRs to test-infra.

Ideally we don't even need to block on a new kube-cross image, we should be able to bump .go-version (or maybe later `toolchain` in the module files) and follow-up with PRs to the images / CI
